File: visualizer/public/javascripts/akkatree.js
$(document).ready(function(){

    var menu = [
        {
            title: 'Collapse',
            action: function(elm, d, i) {
                collapse(d)
            }
        },
        {
            title: 'Expand',
            action: function(elm, d, i) {
                expand(d)
            }
        }
    ]

    //TODO: Remove this ugly id mutable state. Need a monotonic incrementing counter
    var id = 1;
    var root = createRoot(window.innerWidth, window.innerHeight)
    var force = createForceLayout(window.innerWidth, window.innerHeight)
    var vis = graphArea(window.innerWidth, window.innerHeight)

    var eventSource = new EventSource("/events");
    eventSource.onmessage = function(event) {
        akkatree_onmessage(JSON.parse(event.data))
    };
    eventSource.onerror = function(alert) {
        console.log("alert: ", alert)
    }



    function findElementInArray(array, elem) {
        var result = $.grep(array, function(e) { return e.name == elem; })
        if(result.length == 1) {
            return result[0]
        } else {
            return undefined
        }
    }

    //uses isNodeCollapsed property to return appropriate array
    function children(node) {
        if(node.isNodeCollapsed)
            return node.collapsed_children;
        else
            return node.children;
    }

    function insert(path, parent, level) {
        if (path.length == 0) { return; }
        var elem = path.shift();
        var node;
        if (children(parent)) {
            node = findElementInArray(children(parent), elem);
        }
        if (!node) {
            node = {
                "name" : elem,
                "size": 1,
                "id": id++,
                "level" : level,
                "children" : [],
                "collapsed_children" : [],
                "isNodeCollapsed" : false
            };
            children(parent).push(node);
        }
        insert(path, node, level + 1);
    }

    function remove(path, parent){

        var parent_ = parent;

        while (path.length > 1) {
            var elem = path.shift();
            if (parent_ && children(parent_)) {
                parent_ = findElementInArray(children(parent_), elem);
            }
        }

        if (parent_ && children(parent_)) {
            var elem = findElementInArray(children(parent_), path[0]);
            if (elem) {
                var index = children(parent_).indexOf(elem);
                if (index > -1) {
                    children(parent_).splice(index, 1);
                }
            }
        }
    }

    function akkatree_onmessage(msg) {
        showDialogInfo(msg)
        //replacing akka protocol and actor system name with hostname
        var path = msg.actorpath.replace(/akka:\/\/[^\/]+/, msg.host).split("/");
        if (msg.event.type == "started") {
            insert(path, root, 0);
        }

        if (msg.event.type == "terminated") {
            if(path[path.length -1] == "user") { //user actor is terminated, kill the host node
                path.pop() //removes the last element to get to the host path
                remove(path, root);
            } else {
                remove(path, root);
            }
        }
        update();
    }

    function showDialogInfo(msg) {
        $("#dialog-path").html("<h4>Path</h4>" + msg.actorpath);
        $("#dialog-host").html("<h4>Host</h4>" + msg.host);
        $("#dialog-event").html("<h4>Event</h4>" + msg.event.type);
    }

    function update() {
        var nodes = flatten(root),
            links = d3.layout.tree().links(nodes);

        // Restart the force layout.
        force
            .nodes(nodes)
            .links(links)
            .start();

        // Update the links…
        var link = vis.selectAll(".link")
                 .data(links, function(d) { return d.target.id; });

        // Enter any new links.
        link.enter().insert("line", ".node")
            .attr("class", "link")
            .attr("x1", function(d) { return d.source.x; })
            .attr("y1", function(d) { return d.source.y; })
            .attr("x2", function(d) { return d.target.x; })
            .attr("y2", function(d) { return d.target.y; });

        // Exit any old links.
        link.exit().remove();

        // Update the nodes…
        var node = vis.selectAll(".node")
            .data(nodes, function(d) { return d.id; })

        // Exit any old nodes.
        node.exit().remove();

        var nodeEnter = node.enter().append("g")
            .attr("class", "node")
            .on("click", click)
            .call(force.drag);

        nodeEnter.append("circle")
            .attr("r", function(d) { return Math.sqrt((d.size + 1) * 100); })
            .on('contextmenu', d3.contextMenu(menu));

        nodeEnter.append("text")
            .attr("dy", "-1.75em")
            .text(function(d) { return d.name; });

        node.select("circle")
            .style("fill", color);

        force.on("tick", tick(link, node))

    }

    function tick(link, node) {
        return function() {
            link.attr("x1", function(d) { return d.source.x; })
                .attr("y1", function(d) { return d.source.y; })
                .attr("x2", function(d) { return d.target.x; })
                .attr("y2", function(d) { return d.target.y; });
            node.attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; });
        }
    }

    function color(d) {
        if(d.isNodeCollapsed) { return "#ff0000"; }

        var colors = ["#1d4d70", "#3182bd", "#c6dbef", "#ffffff"];
        return colors[d.level % colors.length];
    }

    //used to pin/unpin nodes. This allows to fix part of the tree from moving
    function click(d) {
        if (d3.event.defaultPrevented) return; // ignore drag
        d.fixed = !d.fixed
    }

// Returns a list of all nodes under the root.
    function flatten(root) {
        var nodes = [];

        function recurse(node) {
            if (node.children) node.children.reduce(function(p, v) { return p + recurse(v); }, 0);
            nodes.push(node);
            return node.size;
        }

        recurse(root);
        return nodes;
    }

    function collapse(d) {
        if(d.isNodeCollapsed) { return;} //already collapsed
        d.isNodeCollapsed = true;
        d.collapsed_children = d.children;
        d.children = [];
        update()
    }

    function expand(d) {
        if(!d.isNodeCollapsed) { return; }

        d.isNodeCollapsed = false;
        d.children = d.collapsed_children;
        d.collapsed_children = [];
        update()
    }

    function createRoot(w, h) {
        var root = {
            "name": "akka-tree",
            "size": 0,
            "id" : 0,
            "children" : [],
            "collapsed_children" : [],
            "isNodeCollapsed" : false
        };

        root.fixed = true;
        root.x = w / 2;
        root.y = h / 2;
        return root
    }

    function createForceLayout(w, h) {
       return d3.layout.force()
            .charge(function(d) { return -500; })
            .linkDistance(function(d) { return 50; })
            .size([w - 100, h - 100]);
    }

    function graphArea(w, h) {
        return d3.select("#canvas").append("svg:svg")
            .attr("width", w)
            .attr("height", h);
    }
})

File: visualizer-client/src/main/scala/nworks/js/AkkaTree.scala
package nworks.js

import org.scalajs.d3.D3.Selection
import org.scalajs.d3.Layout.{ForceLayout, GraphNodeForce, GraphLinkForce}
import org.scalajs.dom.raw.{EventSource, MessageEvent, Event}

import scala.scalajs.js
import org.scalajs.{d3, dom}
import org.scalajs.jquery.jQuery

import scala.scalajs.js.{Function2, JSStringOps, JSApp, RegExp}
import scala.scalajs.js.annotation.JSExport
import upickle._
import JSStringOps._

object AkkaTree extends JSApp {

  val width = dom.innerWidth;
  val height = dom.innerHeight;

  var node: d3.D3.UpdateSelection = _
  var link: d3.D3.UpdateSelection = _
  var id = 1

  val root = createRoot(width, height)

  val force: ForceLayout = d3.d3.layout.force()
                  .on("tick", tick _)
                  .charge(-500)
                  .linkDistance(50)
                  .size(js.Array(width - 100.0, height - 100.0))

  val vis: Selection = d3.d3.select("#canvas").append(jQuery("#canvas"), "svg:svg")
    .attr("width", width)
    .attr("height", height)

  def tick(): Unit = {
    link.attr("x1", (d: GraphLinkForce, index: Double) => d.source.x)
        .attr("y1", (d: GraphLinkForce, index: Double) => d.source.y)
        .attr("x2", (d: GraphLinkForce, index: Double) => d.target.x)
        .attr("y2", (d: GraphLinkForce, index: Double) => d.target.y)

    node.attr("transform", (d: GraphNodeForce, index: Double) => "translate(" + d.x + "," + d.y + ")")
  }

  def makeNode(node: AkkaTreeNode): GraphNodeForce = {
    val g = new GraphNodeForce()
    g.index = node.id
    g.fixed = node.fixed
    g.x = node.x
    g.y = node.y
    g
  }
  def flattenNodes(node: AkkaTreeNode, accum: js.Array[GraphNodeForce] = new js.Array[GraphNodeForce]()): js.Array[GraphNodeForce] = {

    for(child <- node.children) {
      flattenNodes(child, accum :+ makeNode(child))
    }
    accum :+ makeNode(node)
  }

  def makeLinks(node: AkkaTreeNode, accum: js.Array[GraphLinkForce] = new js.Array[GraphLinkForce]()): js.Array[GraphLinkForce] = {
    for(child <- node.children) {
      val link = new GraphLinkForce()
      link.source = makeNode(node)
      link.target = makeNode(child)
      makeLinks(child, accum :+ link)
    }
    accum
  }



  def update(force: ForceLayout): Unit = {
    val nodes: js.Array[GraphNodeForce] = flattenNodes(root)
    val links = makeLinks(root)
    force
      .nodes(nodes)
      .links(links)
      .start()

      link = vis.selectAll(".link")
                .data(links){ (a: GraphLinkForce, index: Double) => a.target.index }

      link.enter().insert("line", ".node")
                .attr("x1", (d: GraphLinkForce, index: Double) => d.source.x)
                .attr("y1", (d: GraphLinkForce, index: Double) => d.source.y)
                .attr("x2", (d: GraphLinkForce, index: Double) => d.target.x)
                .attr("y2", (d: GraphLinkForce, index: Double) => d.target.y)

      link.exit().remove()

      node = vis.selectAll(".node")
                 .data(nodes) {(a: GraphNodeForce, index: Double) => a.index }

      // Exit any old nodes.
      node.exit().remove();

      val nodeEnter = node.enter().append("g")
        .attr("class", "node")
//        .on("click", click)
        .call(() => force.drag);

    nodeEnter.append(nodeEnter, "circle")
      .attr("r", 10)
      //.on('contextmenu', d3.contextMenu(menu));

    nodeEnter.append(nodeEnter, "text")
      .attr("dy", "-1.75em")
      .text((d: AkkaTreeNode, index: Double) => d.name)

    node.select("circle")
      .style[AkkaTreeNode]("fill", (d: AkkaTreeNode, index: Double) => color(d));

  }

  def color(d: AkkaTreeNode): String = {
    if(d.isNodeCollapsed) { return "#ff0000"; }

    val colors = Seq("#1d4d70", "#3182bd", "#c6dbef", "#ffffff");
    return colors(d.level.intValue() % colors.length);
  }

  def insert(path: Array[String], parent: AkkaTreeNode, actorpath: String, level: Int): Unit = {
    if(path.length != 0) {
      var elem = path.head;
      val existingNode = children(parent).filter(_.name == elem).headOption
      if(existingNode.isEmpty) {
        id = id + 1
        val node = new AkkaTreeNode(name = elem, size = 1, id = id, level = level)
        children(parent) += node
        insert(path.tail, node, actorpath, level + 1)
      } else {
        insert(path.tail, existingNode.get, actorpath, level + 1)
      }
    }
  }

  //uses isNodeCollapsed property to return appropriate array
  def children(node: AkkaTreeNode): js.Array[AkkaTreeNode] = {
    if(node.isNodeCollapsed)
      return node.collapsedChildren;
    else
      return node.children;
  }

  def remove(path: Array[String], parent: AkkaTreeNode): Unit = {
     def findParent(path: Array[String], node: Option[AkkaTreeNode]): Option[AkkaTreeNode] = {
       if(path.length == 1) node
       else {
         val nextNode = node.flatMap(n => children(n).filter(_.name == path.head).headOption)
         findParent(path.tail, nextNode)
       }
     }

    val immediateParent = findParent(path, Some(parent))
    immediateParent.foreach { p =>
      val elemToRemove = children(p).filter(_.name == path.last).headOption
      elemToRemove.foreach(children(p) -= _)
    }
  }

  def onMessage(msg: AkkaTreeEvent) = {
    showDialogInfo(msg)

    val regex = new RegExp("""akka://[^\/]+""")
    val path: Array[String] = msg.actorpath.jsReplace(regex, msg.host).split("/")
    if(msg.isStarted()) {
      println("found started event ")
      insert(path, root, msg.actorpath, 0);
    }
    if(msg.isTerminated()) {
      println("found terminated event ")
      if(path(path.length - 1) == "user") { //user actor is terminated, kill the host node
        remove(path.init, root)//removes the last element to get to the host path
      } else {
        remove(path, root)
      }
    }

    update(force)
  }

  @JSExport
  override def main(): Unit = {
    val es = new EventSource("/events")
    es.onmessage = {(event: MessageEvent) =>
      val json = event.data.toString
      val akkaTreeEvent: AkkaTreeEvent = read[AkkaTreeEvent](json)
      onMessage(akkaTreeEvent)
    }
    es.onerror = { (error: Event) =>
      println("alert: " + error)
    }
  }

  def createRoot(width: Int, height: Int): AkkaTreeNode = {
    val root = new AkkaTreeNode(name = "akka-tree", size = 0, id = 0, fixed = true, x = width / 2, y = height / 2 )
    root
  }

  def showDialogInfo(msg: AkkaTreeEvent) = {
    jQuery("#dialog-path").html("<h4>Path</h4>" + msg.actorpath);
    jQuery("#dialog-host").html("<h4>Host</h4>" + msg.host);
    jQuery("#dialog-event").html("<h4>Event</h4>" + msg.event);
  }

}
